Skrall/bring feature to exemptions

krallsm opened the pull request:
fixes #687

really wanted this to help us get off of using notScopes for exemptions. I tested various patterns and tried to match it up to the pattern matching in Build-AssignmentDefinitionNode

krallsm commented:
once it's been checked, I'd love to format the script to make it be better formatting wise, but it creates a lot of noise for the PR, so I undid the formatting for my submission

anwather commented:
Apologies I've been away for a couple of weeks - can you provide an update to documentation - and some test cases I can test with? We have some customers utilizing EPAC for thousands of exemptions so I want to make sure we aren't going to break any existing functionality.

anwather commented:
Apologies I've been away for a couple of weeks - can you provide an update to documentation - and some test cases I can test with? We have some customers utilizing EPAC for thousands of exemptions so I want to make sure we aren't going to break any existing functionality.

krallsm commented:
@anwather I added some documentation, if you want to check that out and let me know how it is. It's been a minute, so I'm going to do some additional testing today to make sure all is good, especially focusing on making sure things like the humanReadableName still work as expected along with the subscriptionPatternMatching

krallsm commented:
Tests performed:
To perform a test, I updated my local module with my changes and tested the functionality of the feature by changing the scopes within the exemptions.jsonc file used for providing Policy Exemptions

An example of a test exemption:

```json
{
            "name": "Test-Exemption",
            "displayName": "Test Exemption",
            "description": "This is a test exemption for demonstration purposes.",
            "policyAssignmentId": "/providers/microsoft.management/managementgroups/alz/providers/microsoft.authorization/policyassignments/deploy-asc-monitoring",
            "exemptionCategory": "Mitigated",
            "scopes": [
                "/subscriptions/*/resourceGroups/*"
            ]
}
```

The expected result above is that all resource groups are included in the exemption. This performed as expected locally. This might be able to be turned into an actual test which confirms the resource groups equals the total number of new exemptions.

Other tests, modifying the scopes:
Even though this isn't a practical use case, we added a humanReadableName to this exemption to ensure exemptions still process correctly with the humanReadableName
```
"scopes": [
  "humanReadableName:/subscriptions/*/resourceGroups/*"
]
```
This works as expected

Other tests that were successful by changing the scope appropriately:
```
"humanReadableName:/subscriptions/*/resourceGroups/NetworkWatcherRG"
"/subscriptions/*/resourceGroups/NetworkWatcherRG"
"/*/NetworkWatcherRG"
"/subscriptionsPattern/*Management*"
"/subscriptions/subscriptionsPattern/*anagemen*"
```

So far, haven't found anything to break at all. Gonna add this note in the documentation in a sec, but the biggest note is that if you want to exclude subscriptions, you must use subscriptionsPattern when you want to exclude based on the name. If you exclude that piece when excluding, it's looking at the subscriptionId, rather than it's name. This is expected and why the subscriptionsPattern exists.
